core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java

```diff
@@ -801,62 +813,127 @@ private static void addHttp2Scheme(io.netty.handler.codec.http.HttpHeaders in, U
 
-        // Trailers if it does not have :status.
-        if (server && !in.contains(HttpHeaderNames.STATUS)) {
-            for (Entry<AsciiString, String> entry : in) {
-                final AsciiString name = entry.getKey();
-                final String value = entry.getValue();
-                if (name.isEmpty() || isTrailerBlacklisted(name)) {
-                    continue;
-                }
-                out.add(name, value);
-            }
-        } else {
-            in.forEach((BiConsumer<AsciiString, String>) out::add);
-            out.remove(HttpHeaderNames.CONNECTION);
-            out.remove(HttpHeaderNames.TRANSFER_ENCODING);
+        mergeHeadersHttp2(inputHeaders, outputHeaders, !server, isTrailer, false);
+
+        if (!isTrailer) {
+            mergeHeadersHttp2(additionalHeaders, outputHeaders, !server, false, true);
         }
 
-        if (!out.contains(HttpHeaderNames.COOKIE)) {
-            return out;
+        if (server && endStream) {
+            mergeHeadersHttp2(additionalTrailers, outputHeaders, false, isTrailer, true);
+        }
+
+        outputHeaders.remove(HttpHeaderNames.CONNECTION);
+        outputHeaders.remove(HttpHeaderNames.TRANSFER_ENCODING);
+
+        if (!outputHeaders.contains(HttpHeaderNames.COOKIE)) {
+            return outputHeaders;
         }
 
         // Split up cookies to allow for better compression.
         // https://tools.ietf.org/html/rfc7540#section-8.1.2.5
-        final List<CharSequence> cookies = out.getAllAndRemove(HttpHeaderNames.COOKIE);
+        final List<CharSequence> cookies = outputHeaders.getAllAndRemove(HttpHeaderNames.COOKIE);
         for (CharSequence c : cookies) {
-            out.add(HttpHeaderNames.COOKIE, COOKIE_SPLITTER.split(c));
+            outputHeaders.add(HttpHeaderNames.COOKIE, COOKIE_SPLITTER.split(c));
         }
 
-        return out;
+        return outputHeaders;
+    }
+
+    private static void mergeHeadersHttp2(HttpHeaders inputHeaders, Http2Headers outputHeaders,
+                                          boolean isRequest, boolean isTrailer, boolean isAdditionalHeaders) {
+        final Set<AsciiString> additionalHeaderBlackList = isRequest ? ADDITIONAL_REQUEST_HEADER_BLACKLIST
+                                                                     : ADDITIONAL_RESPONSE_HEADER_BLACKLIST;
+
+        for (Entry<AsciiString, String> entry : inputHeaders) {
+            final AsciiString name = entry.getKey();
+
+            if (name.isEmpty()) {
+                continue;
+            }
+
+            if (isAdditionalHeaders && additionalHeaderBlackList.contains(name)) {
+                continue;
+            }
+
+            if (isTrailer && isTrailerBlacklisted(name)) {
+                continue;
+            }
+
+            outputHeaders.remove(name);
+            inputHeaders.forEachValue(name, newValue -> outputHeaders.add(name, newValue));
+        }
```

Comment:
`inputHeaders.iterator()` can return multiple entries with same name, so `outputHeaders.remove(name);` and `inputHeaders.forEachValue()` are highly inefficient. How about this:

```java
Set<AsciiString> removed = ...;
for (Entry<AsciiString, String> entry : inputHeaders) {
    ... continue ...

    if (removed.add(name)) {
        outputHeaders.remove(name);
    }
    outputHeaders.add(name, value);
}
```

I have some more ideas for improving this, but I think it's not part of this PR.

Reply:
Thanks, I fixed it.